src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Maintainability/EnumValuesShouldBeMonotonicallyIncreasing.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Numerics;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeQuality.Analyzers.Maintainability
+{
+    using static MicrosoftCodeQualityAnalyzersResources;
+
+    /// <summary>
+    /// CA1510: Flags enum values that are not monotonically increasing.
+    /// </summary>
+    public abstract class EnumValuesShouldBeMonotonicallyIncreasing : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA1510";
+
+        internal static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(EnumValuesShouldBeMonotonicallyIncreasingTitle)),
+            CreateLocalizableResourceString(nameof(EnumValuesShouldBeMonotonicallyIncreasingMessage)),
+            DiagnosticCategory.Maintainability,
+            RuleLevel.Disabled,
+            description: null,
+            isPortedFxCopRule: false,
+            isDataflowRule: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.NamedType);
+        }
+
+        private void AnalyzeSymbol(SymbolAnalysisContext context)
+        {
+            if (context.Symbol is not INamedTypeSymbol { TypeKind: TypeKind.Enum } enumSymbol)
+            {
+                return;
+            }
+
+            var members = enumSymbol.GetMembers();
+            BigInteger? previous = null;
```

Comment:
Not sure if using `BigInteger` is good here. It was the easiest to be able to handle all different possible enum underlying types.